### PR TITLE
Supervised GNNModular tasks

### PR DESCRIPTION
**Pull request by tonydavis629**
# Pull Request Template

## Description

This PR enables supervised tasks in GNNModular, and abstracts out the unsupervised edge prediction loss into models.losses.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by rbharath — COMMENTED** [review]
Did a quick skim and nothing major is jumping out at me, but I will hold off on review until #3339 is merged in

File: deepchem/models/losses.py [hunk 1]
```diff
@@ -909,6 +909,69 @@ def loss(atom_vocab_task_atom_pred: torch.Tensor,
         return loss
 
 
+class EdgePredictionLoss(Loss):
+    """
+    Unsupervised graph edge prediction loss.
```
Comment (rbharath) [inline deepchem/models/losses.py:914]: Can you add some more detail about what this loss computes?

File: deepchem/models/tests/test_gnn.py [hunk 2]
```diff
@@ -54,3 +72,48 @@ def test_GNN_save_reload():
...
+
+    dataset, metric = get_multitask_regression_dataset()
+    model = GNNModular(task="regression", num_tasks=3)
+    model.fit(dataset, nb_epoch=100)
+    scores = model.evaluate(dataset, [metric])
+    assert scores['mean_absolute_error'] < 0.2
+
+
+@pytest.mark.torch
+def test_GNN_multitask_classification():
```
Comment (rbharath) [inline deepchem/models/tests/test_gnn.py:112]: Will you add a save/reload test in a follow-up PR?

File: deepchem/models/torch_models/gnn.py [hunk 3]
```diff
@@ -400,6 +421,23 @@ def _prepare_batch(self, batch):
 
         return inputs, labels, weights
 
+    def default_generator(
+            self,
```
Comment (rbharath) [inline deepchem/models/torch_models/gnn.py:425]: Can you add a docstring here? It should explain why we need a custom default generator here instead of the default

**Review by rbharath — COMMENTED** [review]
A few requests for documentation below and we should be good to go

File: deepchem/models/losses.py (hunk 1 above)
Comment (tonydavis629) [reply deepchem/models/losses.py:914]: Done

**Review by rbharath — APPROVED** [review]
LGTM, feel free to merge in once CI is clear
